scheduler/src/cook/mesos/optimizer.clj

```diff
@@ -0,0 +1,158 @@
+;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+;; See the License for the specific language governing permissions and
+;; limitations under the License.
+;;
+(ns cook.mesos.optimizer
+  (:require [cheshire.core :as cheshire]
+            [chime :refer [chime-at chime-ch]]
+            [clj-http.client :as http]
+            [clj-time.core :as t]
+            [clj-time.coerce :as tc]
+            [clj-time.periodic :as periodic]
+            [clojure.core.async :as async]
+            [clojure.tools.logging :as log]
+            [cook.util :refer [lazy-load-var]]
+            [cook.mesos.util :as util]
+            [com.rpl.specter :as sp]
+            [datomic.api :as d :refer (q)]
+            [schema.core :as s]))
+(def PosNum
+  (s/both s/Num (s/pred pos? 'pos?)))
+
+(def PosInt
+  (s/both s/Int (s/pred pos? 'pos?)))
+
+(def NonNegInt
+  (s/both s/Int (s/pred #(<= 0 %) 'pos?)))
+
+(def TimePeriodMs NonNegInt)
+
+;; Elements missing:
+;;   1. Attributes (chip set, az, ..) -- have a combinitorial problem here
+;;   2. Price
+;;   3. Prob preemption in 10 minutes increments (or something like that
+(def HostInfo
+  {:count NonNegInt
+   :instance-type s/Str
+   :cpus PosNum
+   :mem PosNum
+   (s/optional-key :gpus) PosNum
+   :time-to-start TimePeriodMs})
+
+(defprotocol HostFeed
+  "Protocol defining a service to get information on hosts that can be purchased."
+  (get-available-host-info [this] "Returns a list of host info maps conforming
+                                   to HostInfo schema above"))
+
+(def Schedule
+  {TimePeriodMs {:suggested-matches {HostInfo [s/Uuid]}
+                 :suggested-purchases {HostInfo NonNegInt}}})
+
+(defprotocol Optimizer
+  "Protocol defining a tool to produce a schedule to execute"
+  (produce-schedule [this queue running available host-infos]
+                    "Returns a schedule of what jobs to place on what machines
+                     and what hosts to purchase at different time steps.
+                     Conforms to the Schedule schema above
+
+                     Parameters:
+                     queue -- Ordered list of jobs to run (see cook.mesos.schema job)
+                     running -- Set of tasks running (see cook.mesos.schema instance)
+                     available -- Set of offers outstanding
+                     host-infos -- Host infos from HostFeed"))
+
+(defprotocol ScheduleConsumer
+  "Protocol defining a consumer of the optimizer's schedule.
+   An examples is a component to take the schedules suggested
+   purchases and execute on them"
+  ;;Question to self: Should this be run its own thread or let it kick off a thread
+  (consume-schedule [this schedule]
+                    "Act on the schedule passed in.
+                     It is expected this function has side effects
+                     and it will be run in a separate thread"))
+
+(defn create-dummy-host-feed
```

Comment:
Can we add log statements in each of these factory methods notifying us of which instances are created for each of the components? E.g. this will allow us to debug if we have messed up configurations during development if we did not expect to use any of these methods.

Reply:
Sure will do